lasagne/layers/local.py

```diff
@@ -0,0 +1,198 @@
+        ``False``. If it is set to ``True``, its shape should be
+        ``(num_filters, output_rows, output_columns)`` instead.
+        See :func:`lasagne.utils.create_param` for more information.
+
+    nonlinearity : callable or None
+        The nonlinearity that is applied to the layer activations. If None
+        is provided, the layer will be linear.
+
+    flip_filters : bool (default: True)
+        Whether to flip the filters before multiplying them over the input,
+        similar to a convolution (this is the default), or not to flip them,
+        similar to a correlation.
+
+    channelwise : bool (default: False)
+        If ``False``, each filter interacts will all of the input channels as
+        in a convolution. If ``True``, each filter only interacts with the
+        corresponding input channel. That is, each output channel only depends
+        on its filter and on the input channel at the same channel index.
+        In this case, the number of output channels (i.e. number of filters)
+        should be equal to the number of input channels.
+
+    **kwargs
+        Any additional keyword arguments are passed to the `Layer` superclass.
+
+    Attributes
+    ----------
+    W : Theano shared variable or expression
+        Variable or expression representing the filter weights.
+
+    b : Theano shared variable or expression
+        Variable or expression representing the biases.
+
+    Notes
+    -----
+    This implementation computes the output tensor by iterating over the filter
+    weights and multiplying them with shifted versions of the input tensor.
+    This implementation assumes no stride, 'same' padding and no dilation.
+
+    Raises
+    ------
+    ValueError
+        When ``channelwise`` is set to ``True`` and the number of filters
+        differs from the number of input channels, a `ValueError` is raised.
+    """
+    def __init__(self, incoming, num_filters, filter_size, stride=(1, 1),
+                 pad='same', untie_biases=False,
+                 W=init.GlorotUniform(), b=init.Constant(0.),
+                 nonlinearity=nonlinearities.rectify, flip_filters=True,
+                 channelwise=False, **kwargs):
+        self.channelwise = channelwise
+        super(LocallyConnected2DLayer, self).__init__(
+            incoming, num_filters, filter_size, stride=stride, pad=pad,
+            untie_biases=untie_biases, W=W, b=b, nonlinearity=nonlinearity,
+            flip_filters=flip_filters, **kwargs)
+        # require no stride
+        if self.stride != (1, 1):
+            raise NotImplementedError(
+                "LocallyConnected2DLayer requires stride=1 / (1, 1), but got "
+                "%r." % (stride,))
+        # require same convolution
+        if self.pad != 'same':
+            raise NotImplementedError(
+                "LocallyConnected2DLayer requires pad='same', but got %r." %
+                (pad,))
+
+    def get_W_shape(self):
```

Comment:
Would be cool to have a check against variable input shape here. Similar to a dense layer, we require the input shape to be known in a locally-connected layer. Something like:
```python
if any(s is None for s in self.input_shape[1:]):
    raise ValueError(
            "A LocallyConnected2DLayer requires a fixed input shape "
            "(except for the batch size). Got %r." % (self.input_shape,))
```
Otherwise, there'll be an error in creating the weight matrix that's more difficult to interpret.